src/transformers/models/distilbert/modeling_distilbert.py

```diff
@@ -442,6 +442,67 @@ def __init__(self, config):
 
         self.init_weights()
 
+    def get_position_embeddings(self) -> nn.Embedding:
+        """
+        Returns the position embeddings matrix
+        """
+        return self.embeddings.position_embeddings
+
+    def resize_position_embeddings(self, new_num_position_embeddings: int):
+        """
+        Resizes position embeddings matrix of the model if :obj:`new_num_position_embeddings !=
+        config.max_position_embeddings`.
+
+        Arguments:
+            new_num_position_embeddings (:obj:`int`, `optional`):
+                The number of new position embedding matrix. If position embeddings are learned, increasing the size
+                will add newly initialized vectors at the end, whereas reducing the size will remove vectors from the
+                end. If position embeddings not learned (*e.g.* sinusoidal position embeddings), increasing the sise
+                will add correct vectors at the end following the position encoding algorithm, whereas reducing the
+                size will remove vectors from the end.
+        """
+        num_position_embeds_diff = new_num_position_embeddings - self.config.max_position_embeddings
+
+        # no resizing needs to be done if the length stays the same
+        if num_position_embeds_diff == 0:
+            return
+
+        logger.info(f"Setting `config.max_position_embeddings={new_num_position_embeddings}`...")
+        self.config.max_position_embeddings = new_num_position_embeddings
+
+        old_position_embeddings_weight = self.embeddings.position_embeddings.weight.clone()
+
+        self.embeddings.position_embeddings = nn.Embedding(self.config.max_position_embeddings, self.config.dim)
+
+        if self.config.sinusoidal_pos_embds:
+
+            if is_deepspeed_zero3_enabled():
+                import deepspeed
+
+                with deepspeed.zero.GatheredParameters(self.embeddings.position_embeddings.weight, modifier_rank=0):
+                    if torch.distributed.get_rank() == 0:
+                        create_sinusoidal_embeddings(
+                            n_pos=self.config.max_position_embeddings,
+                            dim=self.config.dim,
+                            out=self.embeddings.position_embeddings.weight,
+                        )
+            else:
+                create_sinusoidal_embeddings(
+                    n_pos=self.config.max_position_embeddings,
+                    dim=self.config.dim,
+                    out=self.embeddings.position_embeddings.weight,
+                )
```

Comment:
Seeing this code, can't the DeepSpeed logic be moved to `create_sinusoidal_embeddings`?

Reply:
A general comment: embedded Deepspeed context manager logic leads to really ugly code with a lot of repetition, here and in some of the models.

When the code inside the branches is identical most likely we need to propose to convert these context managers (`GatheredParameters` and `zero.Init`) to have `.start()` / `.stop()` versions of the same, which would then allow to remove the context managers and make the code much more streamlined. 

Except in situations like above where the code in each branch isn't exact.

I've been meaning to think of some abstractions - if you have suggestions I'm all ears.